Source/Atomic/UI/UIFinderWindow.cpp

```diff
@@ -0,0 +1,618 @@
+    UIWindow(context, false),
+    finderMode(0),
+    currentPath(),
+    resultPath(),
+    bookmarks(),
+    bookmarkPaths(),
+    newBookmarkPtr(),
+    newFolderPtr(),
+    bookmarksDirty(0)
+{
+    if (createWidget)
+    {
+        widget_ = new TBFinderWindow(target ? target->GetInternalWidget() : 0, TBIDC(id.CString()));
+        widget_->SetDelegate(this);
+        GetSubsystem<UI>()->WrapWidget(this, widget_);
+    }
+}
+
+UIFinderWindow::~UIFinderWindow()
+{
+    SaveBookmarks();
+    
+    if ( !newFolderPtr.Expired())
+        newFolderPtr->UnsubscribeFromAllEvents();
+    if ( !newBookmarkPtr.Expired() )
+       newBookmarkPtr->UnsubscribeFromAllEvents();
+
+}
+
+void UIFinderWindow::FindFile(const String& title, const String& preset, int dimmer, int width, int height)
+{
+    if (!widget_)
+        return;
+
+    if ( ((TBFinderWindow*)widget_)->Show(title.CString(), preset.CString(), dimmer, width, height) )
+    {
+        CreateBookmarks();
+        PresetCurrentPath(preset);
+        UpdateUiPath();
+        UpdateUiList();
+        UpdateUiResult();
+    }
+}
+
+void UIFinderWindow::FindPath(const String& title, const String& preset, int dimmer, int width, int height)
+{
+    if (!widget_)
+        return;
+
+    finderMode = 1;
+    
+    if ( ((TBFinderWindow*)widget_)->Show(title.CString(), preset.CString(), dimmer, width, height) )
+    {
+        UIWidget *reswid = GetResultWidget();
+        if (reswid) reswid->SetVisibility( UI_WIDGET_VISIBILITY_INVISIBLE );
+        CreateBookmarks();
+        PresetCurrentPath(preset);
+        UpdateUiPath();
+        UpdateUiList();
+        UpdateUiResult();
+    }
+}
+
+bool UIFinderWindow::OnEvent(const tb::TBWidgetEvent &ev)
+{
+
+    if ( ev.type == EVENT_TYPE_CHANGED && ev.target && ((uint32)ev.target->GetID()) == 1 ) 
+    {
+        FileSystem* filesystem = GetSubsystem<FileSystem>();
+        UIWidget *pathwidget = GetPathWidget();  // paste or type in currentpath widget
+        if(pathwidget)
+        {
+            if ( filesystem->DirExists (pathwidget->GetText())) 
+            {
+                if ( pathwidget->GetText() != currentPath ) 
+                {
+                    resultPath = "";
+                    SetCurrentPath (pathwidget->GetText());
+                    UpdateUiList();
+                    UpdateUiResult();
+                }
+            }
+        }
+        return true;
+    }
+
+   if ( ev.type == EVENT_TYPE_POINTER_UP && ev.target && ((uint32)ev.target->GetID()) == 2 ) // go up
+    {
+        GoFolderUp();
+        return true;
+    }
+
+    if ( ev.type == EVENT_TYPE_POINTER_UP && ev.target && ((uint32)ev.target->GetID()) == 3 ) // add bookmark request
+    {
+        // this check is necessary because you can kill the bookmark window with the "X" and
+        // this kills the UI but not the newBookmarkPtr to it, and it doesnt get cleaned up properly.
+        if ( newBookmarkPtr.NotNull() )
```

Comment:
the TB Window has multiple problems, this closing by corner "X" without allowing the application to know being one of them. Others are, there is no concept of topmost window, so you can easily bury windows, and the title area has no provisions to set the size, so on mobile it can be really small.